Implement version 1 of maxonbank

xtyrrell opened the pull request:
Here's Maxonbank v1! Please be harsh in your review! 😄

xtyrrell committed:
chore: add files

xtyrrell commented on src/main/kotlin/com/xtyrrell/maxonbank/coreapi/Money.kt:
```diff
@@ -0,0 +1,14 @@
+ * - precisely stores cents, at least to two decimal places, and possibly more
+ *      depending on the requirements of regulation (eg to support
+ *      calculations on interest we might need to store amounts to eg 5
+ *      decimal places).
```
This would also store curreny info and support multiple currencies

Altusvdm commented on src/main/kotlin/com/xtyrrell/maxonbank/command/Account.kt:
```diff
@@ -0,0 +1,68 @@
+@Aggregate
+class Account() {
+
+    // Question: Is my use of lateinit here correct? I am using it because
```
`lateinit` for anything that gets set in the constructor `@CommandHandler` I think is fine. Axon is forcing you to have an empty constructor by default, but it will always set these values when handling the first event when "reinflating", so dealing with them as nullables is a bit annoying

Altusvdm commented on src/main/kotlin/com/xtyrrell/maxonbank/command/Account.kt:
```diff
@@ -0,0 +1,68 @@
+
+    @CommandHandler
+    constructor(command: OpenAccountCommand) : this() {
+        val aggregateId = UUID.randomUUID()
```
You generally actually want to pass in the Id of an aggregate (I know this is very weird), but the reason being is so that whoever sent this command can look it up.

Currently somewhere you are going `CommandGateway.send(OpenAccountCommand)`, and you can wait for the success of that command, but you don't know what AccountId it created. So you don't know how to send that back to the client. 

You could generate the ID on the API level, so that you know how to respond to the request with an ID, however then you have an issue if the WebRequest times out, did you finish creating the account or not? Letting the client pass in the ID, means the client could request the state of the account on a timeout, and see if it was created or not.

xtyrrell replied:
Okay cool! That's a bit unintuitive but your explanation makes total sense 👌

Altusvdm commented on src/main/kotlin/com/xtyrrell/maxonbank/command/Account.kt:
```diff
@@ -0,0 +1,68 @@
+    private var balance = 0
+
+    @CommandHandler
+    constructor(command: OpenAccountCommand) : this() {
```
You don't need the `: this()`

xtyrrell replied:
If I remove that IntelliJ highlights `constructor(command: OpenAccountCommand)` red with message "Primary constructor call expected" and suggests adding it. Am I doing something wrong?

Altusvdm commented on src/main/kotlin/com/xtyrrell/maxonbank/command/Account.kt:
```diff
@@ -0,0 +1,68 @@
+            throw FundsDepositException("The amount you want to deposit cannot be zero or negative")
+        }
+
+        apply(FundsDepositedEvent(accountId, command.amount))
```
You might want to set an ID on the `FundsDepositedEvent` and you probably want to have that ID passed in along with the command. You wouldn't do anything with that ID in this aggregate, but your query handler would probably need something like `queryDeposit(depositId, accountId)`. This is out of scope for the demo project, but it's important to think about future use cases of your events as once you write them, adding more data to them is very hard.

Same for `FundsWithdrawnEvent`

xtyrrell replied:
Great point!

Altusvdm commented on src/main/kotlin/com/xtyrrell/maxonbank/query/AccountProjector.kt:
```diff
@@ -0,0 +1,95 @@
+        val entry = LedgerEntryView(accountView, entryId, LedgerEntryType.DEPOSIT, event.amount)
+        accountView.addLedgerEntry(entry)
+
+        accountView.balance += event.amount
```
Not sure how this actually works, but I see a possible race condition here. If you handle two `FundsDepositedEvent` s, at the same time. You might possibly load the account twice (with the same initial balance), then add the balance of each indivdual event, and then save again, and thus only account for one of the deposits.

xtyrrell replied:
Well spotted! That's certainly worth some research. I think we can avoid race conditions by using the concurrency mechanisms JPA provides.

xtyrrell replied:
After thinking about this more:

There would be a race condition if
- two different JVMs each have their own AccountProjector instance handling a given `aggregateId`, and one instance receives the first FundsDepositedEvent and the other instance receives the other FundsDepositedEvent;
- or, if one JVM has two AccountProjector instances handling a given `aggregateId`, and one event is sent to each;
- or, if one JVM has one AccountProjector instance handling a given `aggregateId`, and both events are routed to it, but it doesn't process events atomically and so the first half of the FundsDepositedEvent event handler method is run for the first event, then before that can finish, the entire event handler is run for the second event, and only then the event handler for the first event finishes.

The last option seems unlikely. As far as I understand, Axon guarantees that there will always only be one instance for a given `aggregateId`, even in a distributed environment with multiple machines.

Definitely worth more research but so far it seems like this should be fine.

Altusvdm commented on src/main/kotlin/com/xtyrrell/maxonbank/query/AccountProjector.kt:
```diff
@@ -0,0 +1,95 @@
+    }
+
+    @QueryHandler
+    fun handle(query: ListAccountHistoryQuery): List<LedgerEntryView> {
```
You should probably return a nullable list here (like you do for GetAccountAvailableBalanceQuery)

xtyrrell replied:
Ah, nice catch 👌

Altusvdm commented on src/main/kotlin/com/xtyrrell/maxonbank/query/AccountProjector.kt:
```diff
@@ -0,0 +1,95 @@
+    }
+
+    @QueryHandler
+    fun handle(query: GetAccountDetailsQuery): Optional<AccountView> {
```
Where is this `Optional` from? Maybe rather return a Kotlin nullable ie `AccountView?`

Altusvdm commented on src/main/kotlin/com/xtyrrell/maxonbank/query/BankInfoProjector.kt:
```diff
@@ -0,0 +1,26 @@
+import org.springframework.stereotype.Component
+
+@Component
+class BankInfoProjector {
```
I've not used one of these before. I assume what you are trying to do is listen for ANY deposit/withdrawn events, and keep a total running balance for bank across all accounts.

Seems like this only keeps the balance in memory, what happens if your app restarts? Aggregates work by replaying all the events in an aggregate everytime there's a new command (or using snapshots). Presumably everytime your app restarts here, the totalBalance will go to 0 and it will only account for new events (not past ones).

xtyrrell replied:
As far as I understand, Axon should always replay the event log on startup. So I think this should be safe and `totalBalance` should be correct across restarts. But great catch, definitely best to research and confirm properly first.

xtyrrell replied:
It feelss super clean to seperate out event listeners into different classes based on their responsibilitie -- what do you think about having `BankInfoProjector` seperate from `AccountProjector`?

Altusvdm replied:
I think an event can only belong to one aggregate, and an aggregate only "hears" the events that have the same aggregateIdentifier. So you won't be able to have an AccountAggregate (that EventSources the account level balance), and an BankAggragate (that listens to the events from AccountAggregates and maintains an overall balance).

You should be able to setup event listeners, that tracks how far it's listened, and only listens to new events, and then just ensure that everytime your component (BankProjector) goes into memory it starts tracking from the start again. https://docs.axoniq.io/reference-guide/axon-framework/events/event-processors

I haven't really tried to do this, but I like it as a concept, going to do some experimenting. You would probably want to snapshot though, so you don't have to replay every event for every account every time your app restarts.

Keeping the Projectors separate is definitely a great idea, it lets you build fit for purpose queries. It's another thing that feels "wrong" when coming from a traditional non CQRS background, because you often end up duplicating data (maybe saving a ledger entry in AccountProjector and BankInfoProjector), but the info you save in each one is based on the audience that is going to consume it, like maybe in BankInfoProjector you save some debugging info for internal use like what service initiated this transfer, but in AccountProjector you don't need that because an end user would never see it.

xtyrrell replied:
Great, thanks for the feedback :)

Altusvdm commented on src/main/kotlin/com/xtyrrell/maxonbank/web/AccountsController.kt:
```diff
@@ -0,0 +1,77 @@
+        @PathVariable("accountId") accountId: String,
+        @RequestBody depositFundsRequest: DepositFundsRequest
+    ): CompletableFuture<UUID> {
+        // TODO: This returns nothing. Why is that? How can we make it return something, eg a UUID
```
To return something, you have to do the command, then use your query model to return it. (however note your query model is only eventually consistent), so a better solution would be to return the things you know about, ie force the API to take in amount, accountId, ledgerId, and return them.

You can WAIT for a command to finish (and catch any errors), see: https://docs.axoniq.io/reference-guide/extensions/kotlin (onSuccess and onError callbacks)

xtyrrell replied:
Very helpful thank you! I'll apply the "client includes ID in request" pattern here

Altusvdm commented on src/test/kotlin/com/xtyrrell/maxonbank/command/AccountTests.kt:
```diff
@@ -0,0 +1,116 @@
+        }
+
+        @Test
+        fun `should prevent depositing funds for zero or negative amount`() {
```
I would split these two tests, and always only have one call the fixture per test. This makes it much easier to debug what broke when one of the tests starts failing (and ensures your `BeforeEach` is run before each call to fixture)

Altusvdm reviewed (commented):
Looking good!

xtyrrell committed:
refactor: pass entity IDs around explicitly

This commit also renames Commands, Events and Queries and namespaces them for clarity. It also includes minor fixes for issues raised in #1.
